instrumentation/kafka-clients/kafka-clients-0.11/library/src/main/java/io/opentelemetry/instrumentation/kafkaclients/KafkaTracing.java

```diff
@@ -0,0 +1,118 @@
+
+  private static final TextMapSetter<Headers> SETTER = new KafkaHeadersSetter();
+
+  private Instrumenter<ProducerRecord<?, ?>, Void> producerInstrumenter =
+      KafkaSingletons.producerInstrumenter();
+  private Instrumenter<ConsumerRecord<?, ?>, Void> consumerInstrumenter =
+      KafkaSingletons.consumerProcessInstrumenter();
+
+  KafkaTracing() {}
+
+  public static KafkaTracingBuilder newBuilder(OpenTelemetry openTelemetry) {
+    return new KafkaTracingBuilder(openTelemetry);
+  }
+
+  private static TextMapPropagator propagator() {
+    return GlobalOpenTelemetry.getPropagators().getTextMapPropagator();
+  }
+
+  void setProducerInstrumenter(Instrumenter<ProducerRecord<?, ?>, Void> producerInstrumenter) {
+    this.producerInstrumenter = Objects.requireNonNull(producerInstrumenter);
+  }
+
+  void setConsumerInstrumenter(Instrumenter<ConsumerRecord<?, ?>, Void> consumerInstrumenter) {
+    this.consumerInstrumenter = Objects.requireNonNull(consumerInstrumenter);
+  }
+
+  /**
+   * Extract Context from record headers.
+   *
+   * @param headers record headers
+   * @return context
+   */
+  private static Context extractContext(Headers headers) {
+    return propagator().extract(Context.current(), headers, GETTER);
+  }
+
+  /**
+   * Inject current Context to record headers.
+   *
+   * @param headers record headers
+   */
+  private static void inject(Headers headers) {
+    inject(Context.current(), headers);
+  }
+
+  /**
+   * Inject Context to record headers.
+   *
+   * @param context the context
+   * @param headers record headers
+   */
+  private static void inject(Context context, Headers headers) {
+    propagator().inject(context, headers, SETTER);
+  }
+
+  /**
+   * Build and inject span into record. Return Runnable handle to end the current span.
+   *
+   * @param record the producer record to inject span info.
+   * @return runnable to close the current span
+   */
+  public <K, V> Runnable buildAndInjectSpan(ProducerRecord<K, V> record) {
+    Context spanContext = extractContext(record.headers());
+
+    Context current = producerInstrumenter.start(spanContext, record);
+    try (Scope ignored = current.makeCurrent()) {
+      try {
+        inject(record.headers());
+      } catch (Throwable t) {
+        // it can happen if headers are read only (when record is sent second time)
+        logger.error("failed to inject span context. sending record second time?", t);
+      }
+    }
+
+    return () -> producerInstrumenter.end(current, record, null, null);
+  }
+
+  public <K, V> void buildAndFinishChildSpan(ConsumerRecord<K, V> record) {
+    Context parentContext = extractContext(record.headers());
+
+    Context current = consumerInstrumenter.start(parentContext, record);
+    consumerInstrumenter.end(current, record, null, null);
+
+    // Inject created span context into record headers for extraction by client to continue span
+    // chain
+    inject(current, record.headers()); // TODO -- OK?
```

Comment:
I see, this is supposed to propagate context to Vertx handler... I don't think we modify the incoming message this way in any other instrumentation in this repo, this seems somewhat hacky.

Reply:
This is an OpenTracing leftover ... this is how it was done there ...

Reply:
I believe we should remove this code; this is a rather hacky way of passing context; and it won't even work unless there is something other that instruments kafka.